### DISCUSSION TITLE
Exclude a computed field

### DISCUSSION THREAD

**Discussion by orfisko** [pydantic/pydantic#7114] [Pydantic V2]
Does someone know how to exclude a computed field from model_dump() in the model definition? 
The only option I see now is writing my own model_serializer but I feel there must be a better way?

**Comment by snovik75** [comment]
I face the same issue and I think it is a bug. Computed fields should not be serialized if the repr on the field is set to False.

**Comment by Viicos** [comment]
What is the point of excluding a computed field from the serialization output? If you don't want it to be included, you can use a simple `@property`. Or maybe you want to exclude it when using `your_model_instance.model_dump(exclude="your_computed_field")`? The following works:

```python
from pydantic import BaseModel, computed_field

class Model(BaseModel):
    a: int

    @computed_field
    @property
    def b(self) -> int:
        return self.a + 1

Model(a=1).model_dump(exclude="b")
#> {'a': 1}
```

**Comment by orfisko** [comment]
No, because you cannot use setters in that case.

**Reply by Viicos** [reply]
I've updated my comment. Could you please provide an example of what you are trying to achieve?

**Reply by orfisko** [reply]
Well, if you want to know why your suggestion of using the exclude in the model_dump method does not fly, it could make sense to reread all the discussions of the need to be able to exclude a field during serialization _in the model definition_ instead of putting it in the model_dump or dict() method in v1. The moment you have models containing fields pointing to other models which then again point to other models in their model definition this quickly becomes as nightmare. I fail to understand why computed fields should be treated differently in this respect than any other fields.

**Reply by Viicos** [reply]
If you _always_ want the computed field to be excluded, then only using `@property` is what you want. If you want to use setters, you still can as it is standard python and works with `@property`. See https://github.com/pydantic/pydantic/issues/5827, where the same solutions were given.

If you still want to have the computed field included in the JSON Schema? Or apply validation on the setter (then using pydantic validators is the solution)?

**Reply by orfisko** [reply]
Ok, this is new in V2, in V1 this was not possible. That is excellent news actually, thanks!

**Comment by orfisko** [comment]
@Viicos It would make sense to add this to the documentation though. If we now also solve [this issue](https://github.com/pydantic/pydantic/issues/6696) we would have at last a perfect \@property implementation, I look forward to it!

**Comment by benjaoming** [comment]
I ran into this issue because I had `@property` AFTER `@computed_field`.

correct :heavy_check_mark: 

```
    @computed_field
    @property
    def location(self) -> Tuple[float, float]:
        return self.location_lat, self.location_lng
```

wrong :no_entry_sign: 

```
    @property
    @computed_field
    def location(self) -> Tuple[float, float]:
        return self.location_lat, self.location_lng
```

**Reply by Viicos** [reply]
Hum, seems like you inverted the two examples? Doing `@property` on top of `@computed_field` will indeed not work.

**Reply by benjaoming** [reply]
I'll have to get back to this code - all I know is that swapping them made the `location` field disappear from the `model_dump` output in a test case.

**Reply by benjaoming** [reply]
Right, updated the example, thanks :pray:

**Comment by sailxjx** [comment]
It is useful when you want to save the dumped object into database. But currently there seems no easy way to do this.

**Comment by Viicos** [comment]
This recently came up in https://github.com/pydantic/pydantic/discussions/10625, iirc (because people downvote without giving an explanation), https://github.com/pydantic/pydantic/discussions/7114#discussioncomment-6723528 is not enough because it can be tedious to track every computed field to exclude?

**Reply by 1298se** [reply]
The problem with [comment](https://github.com/pydantic/pydantic/discussions/7114#discussioncomment-6723528) is yeah, it gets a bit gnarly with deeply nested models.

**Reply by benjaoming** [reply]
> people downvote without giving an explanation

I think it gets downvoted because the original question is:

> Does someone know how to exclude a computed field from **model_dump()**

Then the comment starts off:

> What is the point of excluding a computed field from the serialization output?

That does not really address the original question.

I assume that the question is about an API endpoint that includes a field that isn't part of the ORM model. An endpoint may handle both input and output data from that field and serialize it to/from other fields in the ORM model. In my note, which resolved my own issue (but now marked off-topic), I was using `location` as an example of serializing between a latitude and longitude field.

**Reply by Viicos** [reply]
> That does not really address the original question.

Because before providing a solution I'd like to know if this is really needed (as I want to avoid [xy problems](https://xyproblem.info/)). The goal of the `computed_field` decorator is to include a property in `model_dump()`, and the user asks if it is possible to have it _not_ included in `model_dump()`. The simplest answer is to not use `computed_field` then. But if I understand correctly now, the user wants to exclude them only under specific scenarios.

**Reply by youcandanch** [reply]
@Viicos can't speak for anyone else, but yeah, that's the exact scenario I ran into this evening. We have some computed fields that represent materialized object relationships, and we want to include those when accessed via our API, but _exclude_ them when we're dumping out to a persistence layer. That way, we can instantiate a serializer from a cached JSON representation, and always retrieve the object relationships at query time. We probably _could_ just stash them in the persistence layer and overwrite them every time, but it does become somewhat troublesome for deep nesting.

What I ended up doing was creating a subclass of BaseModel that includes a `cacheable_data` property, and I'm just implementing it in each serializer we have and excluding the fields that wouldn't be suitable for caching. I can't say that bothers me too much, and I honestly almost prefer it than having something in Pydantic itself, but wanted to give another example.

**Comment by LuisGMM** [comment]
I have a similar case. I have a pydantic model and I want to have some properties I can cache, but have nothing to do with dumping the model. Just a function that caches itself and I can call it without parenthesis for backwards compatibility.

This is the code:

```python
from __future__ import annotations

import json
from typing import Any, Literal

from pydantic import computed_field
from pydantic.dataclasses import dataclass


@dataclass(slots=True, frozen=True)
class Response:
    request: Request

    url: str
    status_code: int
    body: list[int] | None = None  # bytes
    headers: dict[str, str] | None = None

    @computed_field  # type: ignore[misc]
    @property
    def content(self) -> str | None:
        if self.body is None:
            msg = 'Response body is not available. There is no content to decode.'
            logging.api.log(logging.WARNING, msg)

            return None

        encoding: str | None = None

        if self.headers is None:
            msg = 'Response headers are not available. Cannot determine encoding. Assuming utf-8.'
            logging.api.log(logging.WARNING, msg)

        else:
            content_type = self.headers.get('content-type', None) or self.headers.get('Content-Type', None)

            if content_type is None:
                msg = 'Content-Type header is not available. Cannot determine encoding. Assuming utf-8.'

                logging.api.log(logging.WARNING, msg)

            else:
                content_type = content_type.lower()
                charset = content_type.split('charset=')[-1]

                if charset != content_type:
                    encoding = charset

                elif 'application/json' in content_type:
                    # NOTE: The default encoding for JSON is utf-8 as per RFC 4627.
                    encoding = 'utf-8'

                else:
                    msg = 'Content-Type header does not contain a charset. Cannot determine encoding. Assuming utf-8.'
                    logging.api.log(logging.WARNING, msg)

        if encoding is None:
            encoding = 'utf-8'

        body_bytes = bytes(self.body)

        return body_bytes.decode(encoding)

    @computed_field  # type: ignore[misc]
    @property
    def content_deserialized(self) -> Any | None:
        if self.content is None:
            return None

        try:
            return json.loads(self.content)

        except json.JSONDecodeError:
            msg = 'Response content is not a valid JSON.'
            logging.api.log(logging.WARNING, msg)

            return None

    @computed_field  # type: ignore[misc]
    @property
    def content_json(self) -> dict[str, Any] | None:
        if self.content_deserialized is None:
            return None

        if not isinstance(self.content_deserialized, dict):
            msg = 'Response content is not a JSON map object.'
            logging.api.log(logging.WARNING, msg)

            return None

        return self.content_deserialized

    @computed_field  # type: ignore[misc]
    @property
    def content_list(self) -> list[Any] | None:
        if self.content_deserialized is None:
            return None

        if not isinstance(self.content_deserialized, list):
            msg = 'Response content is not a JSON list object.'
            logging.api.log(logging.WARNING, msg)

            return None

        return self.content_deserialized
```

As it is, it does not as I expected it.

**Comment by gsakkis** [comment]
@Viicos can we reopen this? AFAICT there's no (good) solution for this:

> But if I understand correctly now, the user wants to exclude them only under specific scenarios.

**Comment by Viicos** [comment]
A solution is to use `round_trip=True`:

```python
from pydantic import BaseModel, computed_field

class Model(BaseModel):
    @computed_field
    def test(self) -> int:
        return 1

Model().model_dump()
#> {'test': 1}
Model().model_dump(round_trip=True)
#> {}
```

Note that using `round_trip` has other effects, especially with JSON serialization.

Before blindly including an exclude option, I'd like to fully understand the use case. It was [mentioned](https://github.com/pydantic/pydantic/discussions/7114#discussioncomment-10704276) in this thread:

> It is useful when you want to save the dumped object into database. But currently there seems no easy way to do this.

Apart from having to exclude computed fields, trying to save a Pydantic object in a database might have other implications that aren't relevant when normally dumping a model.

**Reply by hugo-ricateau** [reply]
`@computed_field` also has the beneficial effect to include these fields in the model `repr` (which is really useful for debugging for instance); being able to select on a field-by-field basis, whether to include it in the `.model_dump()` output (e.g. via `@computed_field(exclude=True)`), would be helpful for such use-cases.

`round_trip=True` indeed does the job when all computed fields can be excluded, but unfortunately my use-case requires some "genuine" ones.

**Comment by Viicos** [comment]
https://github.com/pydantic/pydantic/issues/5827 is the issue tracking this.

**Comment by Viicos** [comment]
A `exclude_computed_fields` option is added in https://github.com/pydantic/pydantic/pull/12334, landing in 2.12.

**However**, please consider using `round_trip` if you are trying to achieve round-tripping behavior!

**Comment by luzede** [comment]
As of `pydantic: v2.13.4` there are 2 ways:

1) Exclude all computed fields using `.model_dump` method
   ```python
   model_instance.model_dump(exclude_computed_fields=True)
   ```
2) Exclude certain fields by providing truthy callable to `exclude_if` parameter
   ```python
   @computed_field(exclude_if=lambda _: True)
   ```
